docs/tutorials/Containers.md

```diff
@@ -138,14 +138,31 @@ As the configuration will require more knowledge about your execution environmen
 
 ##### Local environments
 
-On local backends, you have to configure Cromwell to use a different `submit-docker` script that would start Singularity instead of docker. Singularity requires docker images to be prefixed with the prefix `docker://`.
+On local backends, you have to configure Cromwell to use a different 
+`submit-docker` script that would start Singularity instead of docker. 
+Singularity requires docker images to be prefixed with the prefix `docker://`.
 An example submit script for Singularity is:
 ```bash
-singularity exec --bind ${cwd}:${docker_cwd} docker://${docker} ${job_shell} ${script}
+singularity exec --containall --bind ${cwd}:${docker_cwd} docker://${docker} ${job_shell} ${script}
 ```
 
 As the `Singularity exec` command does not emit a job-id, we must include the `run-in-background` tag within the the provider section in addition to the docker-submit script. As Cromwell watches for the existence of the `rc` file, the `run-in-background` option has the caveat that we require the Singularity container to successfully complete, otherwise the workflow might hang indefinitely.
 
+The `--containall` flag is **important**. By default Singularity will:
```

Comment:
Interesting suggestion - don't most users _want_ that seamless transition between host and container, especially for writing files? Is there any reason to not suggest pulling the image first and then running exec against the binary?

Reply:
Mounting local filesystem locations without explicitly recording file inputs and outputs in your WDL makes it harder to determine call cache validity and makes the WDL non-portable.

Reply:
Ah, ok - so the context is when using wdl, and yes that makes sense. Could we add that as a clarification so it's not lost to dinosaurs like me that don't read it with this context?

Reply:
Yeah I agree with this, I think it’s super important for reproducibility To have this containall flag. I’ve seen random things like python installations break inside a container without it.

I think adding a precursor sentence would be good, something like “to ensure reproducibility and an isolated environment inside the container, ‘containall’ is an **important** function ...”

Reply:
> Is there any reason to not suggest pulling the image first and then running exec against the binary?

Yes. @TMIguelT has documented this in this issue: https://github.com/broadinstitute/cromwell/issues/5063.

Using a cache prevents pulling the same image over and over again. To populate the cache a command needs to be run before the submit command.

Reply:
Well actually it probably is ideal to pull the image first (before running cromwell at all), but then it's no longer a download-and-run setup, it requires manual intervention. 

Imagine if you didn't know anything about singularity and WDL at all, and you just wanted to run some WDL you found online, using your pre-configured Cromwell + Singularity. You would have no idea how to work out what containers are used, or even what *containers* are, necessarily, which is why I would like Cromwell to do this for us.

Reply:
If the image is pulled once to a shared filesystem it’s the shared target for all the jobs, like any binary or script. Is the filesystem not shared? What am I missing?

Reply:
@vsoch you are correct. But by using exec we let Singularity do all the work of localizing and storing the images instead of doing it ourselves in bash.

I prefer to let tools work for me, instead if me working for the tools :wink:.

Reply:
Yes so if it's pulled once it will work and be shared across all workers. The problem is that pulling manually requires knowledge of singularity, and pulling automatically inside Cromwell requires a file lock to prevent wacky race conditions where all workers are simultaneously writing to the cache.

Reply:
Well I wouldn't desribe this filelock as a problem. I mean, it solves the problem nicely for the singularity use case. Given the design of `flock` and Cromwell I  am quite certain this effectively solves all problems.

Reply:
Ah understood. For clusters not using Cromwell, there has been a little bit of disaster to not pull first (and having nodes each starting to pull a 500GB container at the same time) so I advocate for puling first. One more question - if we always exec with a particular container URI and the latest (or tag) changes, Singularity will detect a difference in the sha and  pull again. The user might not be aware that the container has changed. Pulling first and interacting with the binary solves this issue.

Reply:
Encouraging users to use the container hash? 😬

Reply:
I think that kind of usage is hard for most, you have to look it up or know how to inspect Docker. If you could figure out the percentage of users that pull with a hash I’d suspect it to be very small. A pull before running a job to produce a binary that you confidently use (and reuse) is still my preference here, although I suspect that’s not the Cromwell way :)

Reply:
@vsoch. This does pull first! It simply uses the exec command to do so, because that way singularity will do the pull and use the cache. It uses a different cache for `exec` than it `uses` for pull (the layers are the same, but with exec it also creates a build image in the cache.)

Reply:
Only AFTER the exec statement is the job submitted.

Reply:
Also @TMiguelT and @vsoch doesn't cromwell use the container hash by default? And stores this in the database?

Also if you are a decent person and use biocontainers instead of rolling your own, you should not get any changing tags, as the build system for biocontainers does not allow this.

Reply:
I don't know how the Cromwell internals work. Certainly this would be ideal.

I also didn't realise biocontainers don't override tags, but still many images from external sources (`miniconda:latest`, or `ubuntu:latest`, for example) risk changing occasionally.

Reply:
> I don't know how the Cromwell internals work. Certainly this would be ideal.

There is quite some code in cromwell to handle docker specifics. Including hash lookup.

> I also didn't realise biocontainers don't override tags, but still many images from external sources (miniconda:latest, or ubuntu:latest, for example) risk changing occasionally.

That is true. In biowdl tasks we only use biocontainers for this reason. A tag is more readable than a hash. For some corner case stuff we use a debian container, but that one is fixed at a specific sha256 hash. 

But anyway, we are digressing :wink:. This does not matter much for this specific PR, I think.

Reply:
It matters only in the sense that this config doesn't guarantee reproducibility, but it's probably not its job to do so.

Reply:
Agreed. But indeed, no config will guarantee reproducibility of the user chooses the wrong containers :confused:.

Reply:
As regards to this, should we also give a mention to the prepull-singularity script from BioWDL? I have written the config so that it is not needed, but maybe users want it is an option? https://github.com/biowdl/prepull-singularity